File: lib/quantize.ts
import { oklabDistanceSquared, oklabToRgb, rgbToOklab, type Oklab } from "./color";
import { mergeSimilarColors } from "./palette-optimizer";
import { mulberry32 } from "./prng";
import { cellRgb, type CellColorBuffer, type RGB } from "./types";

export interface QuantizeResult {
  /** One palette index per cell, row-major, same length as the input grid. */
  cellPaletteIndex: Uint8Array;
  /** Representative RGB color for each palette entry, length k (or fewer if k > distinct colors). */
  palette: RGB[];
}

export interface ColorQuantizer {
  quantize(cells: CellColorBuffer, colorCount: number): QuantizeResult;
}

/**
 * Mean of a cluster's member colors *in OKLab space*, converted back to RGB
 * (with gamut clamping via `oklabToRgb`) -- the correct Lloyd-update
 * centroid for the squared-OKLab-distance objective assignment and ICM/
 * contour-cleanup optimization actually use throughout this pipeline.
 *
 * A linear-RGB mean (this function's predecessor, `meanRgbLinear`) does NOT
 * minimize squared OKLab error for a given membership -- a mean only
 * minimizes squared error in the coordinate system it's computed in, and
 * linear RGB isn't that system here. A domain-expert review (HANDOVER.md
 * D11) had flagged the *need* to recompute post-optimization, but the
 * recompute itself still used a linear-RGB mean, leaving the same
 * inconsistency; the code-review that caught this (2026-09-09, finding 3)
 * reproduced it directly: on a 100x60 grayscale ramp through the default
 * two-color pipeline, recomputing the same final memberships in OKLab
 * reduced mean squared OKLab error by ~6.9% without moving a single stitch,
 * and on a simple 50/50 black/white cluster the reduction was ~26% (0.337 ->
 * 0.250). Linear-light averaging remains the correct approach for the
 * *spatial downsample* (`downsampleToGrid`) -- that's a genuinely different
 * operation (reconstructing what a printed cell's average appearance would
 * be) from *this* one (finding the representative color that best serves the
 * clustering objective already in effect).
 */
export function meanRgbOklab(cells: CellColorBuffer, indices: number[]): RGB {
  let l = 0;
  let a = 0;
  let b = 0;
  for (const i of indices) {
    const [ol, oa, ob] = rgbToOklab(cellRgb(cells, i));
    l += ol;
    a += oa;
    b += ob;
  }
  const n = indices.length || 1;
  return oklabToRgb([l / n, a / n, b / n]);
}

/** Deterministic k-means++ seeding: spreads initial centroids apart instead of picking randomly. */
function kMeansPlusPlusSeeds(oklabColors: Oklab[], k: number, rng: () => number): Oklab[] {
  const seeds: Oklab[] = [oklabColors[Math.floor(rng() * oklabColors.length)]];
  const distSq = new Float64Array(oklabColors.length).fill(Infinity);

  while (seeds.length < k) {
    let total = 0;
    for (let i = 0; i < oklabColors.length; i++) {
      const d = oklabDistanceSquared(oklabColors[i], seeds[seeds.length - 1]);
      if (d < distSq[i]) distSq[i] = d;
      total += distSq[i];
    }
    if (total === 0) {
      // All remaining points coincide with an existing seed; pad with duplicates.
      seeds.push(oklabColors[Math.floor(rng() * oklabColors.length)]);
      continue;
    }
    let threshold = rng() * total;
    let chosen = 0;
    for (let i = 0; i < oklabColors.length; i++) {
      threshold -= distSq[i];
      if (threshold <= 0) {
        chosen = i;
        break;
      }
    }
    seeds.push(oklabColors[chosen]);
  }
  return seeds;
}

const MAX_ITERATIONS = 30;
const CONVERGENCE_THRESHOLD_SQ = 0.0001;

/** Standard Lloyd's-algorithm refinement to convergence from a given set of initial centroids. */
function runLloyd(oklabColors: Oklab[], initialCentroids: Oklab[]): { centroids: Oklab[]; assignments: Uint8Array } {
  let centroids = initialCentroids;
  const assignments = new Uint8Array(oklabColors.length);

  for (let iter = 0; iter < MAX_ITERATIONS; iter++) {
    for (let i = 0; i < oklabColors.length; i++) {
      let best = 0;
      let bestDist = Infinity;
      for (let c = 0; c < centroids.length; c++) {
        const d = oklabDistanceSquared(oklabColors[i], centroids[c]);
        if (d < bestDist) {
          bestDist = d;
          best = c;
        }
      }
      assignments[i] = best;
    }

    const sums = centroids.map(() => [0, 0, 0]);
    const counts = new Array(centroids.length).fill(0);
    for (let i = 0; i < oklabColors.length; i++) {
      const c = assignments[i];
      sums[c][0] += oklabColors[i][0];
      sums[c][1] += oklabColors[i][1];
      sums[c][2] += oklabColors[i][2];
      counts[c]++;
    }

    let maxShiftSq = 0;
    const newCentroids: Oklab[] = centroids.map((old, c) => {
      if (counts[c] === 0) return old;
      const next: Oklab = [sums[c][0] / counts[c], sums[c][1] / counts[c], sums[c][2] / counts[c]];
      maxShiftSq = Math.max(maxShiftSq, oklabDistanceSquared(old, next));
      return next;
    });
    centroids = newCentroids;
    if (maxShiftSq < CONVERGENCE_THRESHOLD_SQ) break;
  }

  return { centroids, assignments };
}

/**
 * Builds the final RGB palette straight from `runLloyd`'s own converged
 * OKLab centroids -- each one already *is* the exact OKLab mean of the
 * cells assigned to it (that's what makes a centroid step a real Lloyd
 * update), so converting it to RGB (via `oklabToRgb`, gamut-clamped) is the
 * correct representative color, not a second, differently-computed mean
 * over the same membership (code-review 2026-09-09, finding 3).
 */
function buildPaletteFromAssignment(
  centroids: Oklab[],
  assignments: Uint8Array
): { cellPaletteIndex: Uint8Array; palette: RGB[] } {
  const counts = new Array(centroids.length).fill(0);
  for (const c of assignments) counts[c]++;

  const remap = new Int16Array(centroids.length).fill(-1);
  const palette: RGB[] = [];
  centroids.forEach((centroid, c) => {
    if (counts[c] === 0) return;
    remap[c] = palette.length;
    palette.push(oklabToRgb(centroid));
  });

  const cellPaletteIndex = new Uint8Array(assignments.length);
  for (let i = 0; i < assignments.length; i++) cellPaletteIndex[i] = remap[assignments[i]];

  return { cellPaletteIndex, palette };
}

// Looser than palette-optimizer.ts's own DEFAULT_MERGE_DISTANCE_SQUARED
// (0.0004, tuned to catch only genuinely-tight near-duplicates for its own
// late-pipeline dedup role): this pass runs immediately after raw
// quantization, specifically looking for palette redundancy worth trading
// away for a real minority color the population-weighted k-means objective
// otherwise never allocates a slot to (HANDOVER.md D19/D20). Deliberately
// still much tighter than a "these look similar to a human" threshold --
// only merges colors close enough that losing the distinction barely
// changes reconstruction error, so genuinely different content (checked
// against the busy-multi-hue case in testing) isn't merged away.
const REINVEST_MERGE_THRESHOLD = 0.012;

/**
 * Redistributes palette budget freed by merging redundant colors to
 * whichever cell is currently the single worst-represented in the whole
 * image (largest reconstruction error to its own assigned color) --
 * repeated once per freed slot. This is the classic split/grow step from
 * LBG-style vector-quantization codebook design (Linde-Buzo-Gray 1980):
 * grow a codebook by always splitting off the currently-worst-served point,
 * not by density or geometry. Reconstruction error is a direct, real
 * measurement of "how badly does an actual color need help," unlike the
 * population/geometry proxies two earlier, reverted attempts relied on
 * (HANDOVER.md D18, D19) -- a genuinely rare, saturated color like a small
 * eye is, by construction, the worst-served point once the rest of the
 * palette has settled onto the dominant content, so it gets first claim on
 * any freed slot without needing to out-compete a majority region's
 * population anywhere in the process.
 */
function injectWorstFitClusters(oklabColors: Oklab[], assignment: Uint8Array, centroids: Oklab[], slotsToAdd: number) {
  const nextAssignment = assignment.slice();
  let nextCentroids = centroids.slice();

  for (let slot = 0; slot < slotsToAdd; slot++) {
    let worstIndex = 0;
    let worstDist = -1;
    for (let i = 0; i < oklabColors.length; i++) {
      const d = oklabDistanceSquared(oklabColors[i], nextCentroids[nextAssignment[i]]);
      if (d > worstDist) {
        worstDist = d;
        worstIndex = i;
      }
    }

    const newCentroid = oklabColors[worstIndex];
    const newClusterIndex = nextCentroids.length;
    nextCentroids = [...nextCentroids, newCentroid];

    for (let i = 0; i < oklabColors.length; i++) {
      const distToNew = oklabDistanceSquared(oklabColors[i], newCentroid);
      const distToCurrent = oklabDistanceSquared(oklabColors[i], nextCentroids[nextAssignment[i]]);
      if (distToNew < distToCurrent) nextAssignment[i] = newClusterIndex;
    }
  }

  return { assignment: nextAssignment, centroids: nextCentroids };
}

/**
 * Plain single-stage k-means in OKLab space (see HANDOVER.md D6 for why
 * OKLab over CIELAB+CIEDE2000). Squared Euclidean distance isn't a
 * compromise here — Lloyd's algorithm's centroid-update step is only valid
 * under that exact metric, and CIEDE2000 isn't even a metric (violates the
 * triangle inequality) — confirmed independently by the domain-expert
 * review, HANDOVER.md D7. Each cell has already been box-averaged (in
 * linear light) by `downsampleToGrid` before reaching here. The reported
 * palette color is `buildPaletteFromAssignment`'s direct `oklabToRgb`
 * conversion of each converged centroid — the correct Lloyd-update
 * centroid for the squared-OKLab objective this quantizer actually
 * minimizes (see `meanRgbOklab`'s docstring above for why a linear-RGB
 * mean would be a different, less accurate color for that same
 * membership). `buildPattern` in `lib/pattern.ts` recomputes this again
 * from each color's *final* post-optimization membership before a chart
 * is rendered, since ICM/contour-cleanup reassign cells after this
 * function returns — this quantizer's own returned palette is only ever
 * final as-is when called directly (e.g. with `optimize: false`, or from
 * a test) rather than through the normal `buildPattern` pipeline.
 *
 * This is "Original" in the app's generation-mode switch (HANDOVER.md D20):
 * exactly the algorithm this project shipped with, before D18/D19/D20's
 * investigation into small-region color loss. Kept available on purpose,
 * not just as a fallback — it has a real, opposite trade-off from
 * `kMeansQuantizer` below (simpler, more population-driven palettes; can
 * miss a small distinct region at low color counts) that some source
 * images and preferences suit better.
 */
export const plainKMeansQuantizer: ColorQuantizer = {
  quantize(cells: CellColorBuffer, colorCount: number): QuantizeResult {
    const cellCount = cells.width * cells.height;
    const k = Math.max(1, Math.min(colorCount, cellCount));
    const oklabColors = new Array<Oklab>(cellCount);
    for (let i = 0; i < cellCount; i++) oklabColors[i] = rgbToOklab(cellRgb(cells, i));

    const rng = mulberry32(0xc0ffee ^ cellCount ^ k);
    const initialSeeds = kMeansPlusPlusSeeds(oklabColors, k, rng);
    const initial = runLloyd(oklabColors, initialSeeds);
    return buildPaletteFromAssignment(initial.centroids, initial.assignments);
  },
};

/**
 * "Latest" in the app's generation-mode switch (HANDOVER.md D20, the
 * default): runs `plainKMeansQuantizer` first, then checks whether any of
 * the resulting colors are redundant enough to merge
 * (`REINVEST_MERGE_THRESHOLD`, looser than `palette-optimizer.ts`'s own
 * late-pipeline dedup threshold) and, if so, reinvests each freed slot into
 * whichever cell is currently worst-served (`injectWorstFitClusters`), then
 * re-converges. This only changes anything when real redundancy is
 * actually found — an image with no redundant colors (verified in testing
 * against a genuinely multi-hued fixture with no dominant majority) takes
 * the exact same path as the plain quantizer above, unlike two earlier,
 * reverted attempts at this same underlying problem (HANDOVER.md D18/D19),
 * which altered every image's clustering unconditionally regardless of
 * whether it needed it. Real trade-off, not a strict improvement: measured
 * confetti/complexity on some ordinary noisy photos rises modestly (still
 * within the project's own regression-suite tolerance bands) in exchange
 * for reliably surfacing a small, real, perceptually-distinct region much
 * sooner — which is exactly why both modes stay selectable rather than one
 * replacing the other outright.
 *
 * `freedSlots` also recovers plain, ordinary Lloyd's-algorithm attrition
 * (a k-means++ seed's Voronoi region going empty during refinement), not
 * just slots `mergeSimilarColors` frees from genuine redundancy (2026-09-11
 * review, HANDOVER.md D39/G-020 M2) -- comparing the merged survivor count
 * against `targetK` (the actual requested/clamped color budget) rather
 * than against `initialResult`'s own count means a color lost to bad-luck
 * seeding gets the same reinvestment chance as one lost to a real
 * redundancy merge, using the exact same, already-proven mechanism. This
 * is provably safe against the "genuinely fewer distinct colors than k"
 * case the collapse behavior above is *supposed* to produce (see
 * `quantize.spec.ts`'s "collapses to the number of distinct colors... never
 * producing empty entries" test): when every cell already sits exactly on
 * its own centroid (zero reconstruction error everywhere, because there's
 * truly nothing left to split), any speculative injected cluster attracts
 * no cells away from its neighbor and comes back empty from `runLloyd`'s
 * own reconvergence pass, so `buildPaletteFromAssignment` drops it again --
 * the same self-correcting property that already makes `injectWorstFitClusters`
 * safe to call unconditionally in the merge-triggered case below. Confirmed
 * against a reproducible real-world case found by brute-force search over
 * random distinct-color fixtures (25 cells / 13 distinct colors, k=5):
 * before this fix, both `plainKMeansQuantizer` and `kMeansQuantizer`
 * silently returned only 4 colors; after, `kMeansQuantizer` recovers the
 * full 5 requested (`quantize.spec.ts`'s attrition-recovery test).
 */
export const kMeansQuantizer: ColorQuantizer = {
  quantize(cells: CellColorBuffer, colorCount: number): QuantizeResult {
    const initialResult = plainKMeansQuantizer.quantize(cells, colorCount);
    const targetK = Math.min(colorCount, cells.width * cells.height);
    if (targetK < 3) {
      // Nothing meaningful to redistribute (k=1/2, or the image only has a
      // couple of real distinct colors) -- skip straight to the plain result.
      return initialResult;
    }

    const merged = mergeSimilarColors(initialResult.cellPaletteIndex, initialResult.palette, REINVEST_MERGE_THRESHOLD);
    const freedSlots = Math.max(0, targetK - merged.palette.length);
    if (freedSlots <= 0) {
      return initialResult;
    }

    const cellCount = cells.width * cells.height;
    const oklabColors = new Array<Oklab>(cellCount);
    for (let i = 0; i < cellCount; i++) oklabColors[i] = rgbToOklab(cellRgb(cells, i));

    const mergedOklab = merged.palette.map(rgbToOklab);
    const injected = injectWorstFitClusters(oklabColors, merged.cellPaletteIndex, mergedOklab, freedSlots);
    const refined = runLloyd(oklabColors, injected.centroids);
    return buildPaletteFromAssignment(refined.centroids, refined.assignments);
  },
};

File: rust/cs-core/src/quantize.rs
//! Port of `lib/pipeline/quantize.ts`: k-means++ seeding, Lloyd with Hamerly bounds (D177), "Original" and "Latest"
//! (merge, reinvest the freed slots in the worst-served cells, re-converge; D20, D39).

use crate::color::{oklab_distance_sq, oklab_to_rgb, rgb_to_oklab, Oklab, Rgb};
use crate::jsmath;
use crate::palette_merge::merge_similar_colors;
use crate::prng::Mulberry32;
use rayon::prelude::*;

/// Points per parallel task.
pub(crate) const CHUNK: usize = 8192;

const MAX_ITERATIONS: usize = 30;
const CONVERGENCE_THRESHOLD_SQ: f64 = 0.0001;
const BOUND_MARGIN: f64 = 1e-9;
pub const REINVEST_MERGE_THRESHOLD: f64 = 0.012;
pub const WORST_FIT_IMPORTANCE_BOOST: f64 = 1.0;

#[derive(Clone, Copy, PartialEq, Eq, Debug)]
pub enum Quantizer {
    Original,
    Latest,
}

#[inline]
fn point_at(points: &[f64], i: usize) -> Oklab {
    [points[i * 3], points[i * 3 + 1], points[i * 3 + 2]]
}

/// `meanOklabAsRgb`.
/// A thread's colour under Vivid (G-062, D212): mean lightness with the chroma of its most colourful quarter,
/// gamut-mapped. Port of `vividOklabAsRgb`.
pub fn vivid_oklab_as_rgb(cell_oklab: &[f64], indices: &[usize]) -> Rgb {
    let mut sum_l = 0.0;
    for &i in indices {
        sum_l += cell_oklab[i * 3];
    }
    // Ordered by chroma, ties by cell index, so both languages take the same quarter.
    let mut by_chroma: Vec<(f64, usize)> = indices
        .iter()
        .map(|&i| {
            let a = cell_oklab[i * 3 + 1];
            let b = cell_oklab[i * 3 + 2];
            (a * a + b * b, i)
        })
        .collect();
    by_chroma.sort_by(|x, y| x.0.partial_cmp(&y.0).unwrap().then(x.1.cmp(&y.1)));
    let from = ((by_chroma.len() as f64) * (1.0 - crate::downsample::VIVID_TOP_SHARE)).floor() as usize;
    let mut sum_a = 0.0;
    let mut sum_b = 0.0;
    for &(_, i) in &by_chroma[from..] {
        sum_a += cell_oklab[i * 3 + 1];
        sum_b += cell_oklab[i * 3 + 2];
    }
    let top = (by_chroma.len() - from) as f64;
    let lin = crate::color::gamut_map_oklab_to_linear(
        sum_l / indices.len() as f64,
        sum_a / top,
        sum_b / top,
    );
    [
        crate::color::linear_to_srgb(lin[0]),
        crate::color::linear_to_srgb(lin[1]),
        crate::color::linear_to_srgb(lin[2]),
    ]
}

pub fn mean_oklab_as_rgb(cell_oklab: &[f64], indices: &[usize]) -> Rgb {
    let (mut l, mut a, mut b) = (0.0, 0.0, 0.0);
    for &i in indices {
        l += cell_oklab[i * 3];
        a += cell_oklab[i * 3 + 1];
        b += cell_oklab[i * 3 + 2];
    }
    let n = if indices.is_empty() {
        1.0
    } else {
        indices.len() as f64
    };
    oklab_to_rgb([l / n, a / n, b / n])
}

fn kmeans_plus_plus_seeds(points: &[f64], k: usize, rng: &mut Mulberry32) -> Vec<Oklab> {
    let n = points.len() / 3;
    let nf = n as f64;
    let mut seeds = vec![point_at(points, (rng.next_f64() * nf).floor() as usize)];
    let mut dist_sq = vec![f64::INFINITY; n];
    while seeds.len() < k {
        let [sl, sa, sb] = *seeds.last().unwrap();
        dist_sq
            .par_chunks_mut(CHUNK)
            .enumerate()
            .for_each(|(c, dist)| {
                for (j, dd) in dist.iter_mut().enumerate() {
                    let i = c * CHUNK + j;
                    let dl = points[i * 3] - sl;
                    let da = points[i * 3 + 1] - sa;
                    let db = points[i * 3 + 2] - sb;
                    let d = dl * dl + da * da + db * db;
                    if d < *dd {
                        *dd = d;
                    }
                }
            });
        // Summed in point order, as the TypeScript does.
        let mut total = 0.0;
        for &d in &dist_sq {
            total += d;
        }
        if total == 0.0 {
            seeds.push(point_at(points, (rng.next_f64() * nf).floor() as usize));
            continue;
        }
        let mut threshold = rng.next_f64() * total;
        let mut chosen = 0;
        for i in 0..n {
            threshold -= dist_sq[i];
            if threshold <= 0.0 {
                chosen = i;
                break;
            }
        }
        seeds.push(point_at(points, chosen));
    }
    seeds
}

struct Scan {
    best: usize,
    best_dist: f64,
    second_dist: f64,
}

#[inline]
fn scan_point(points: &[f64], i: usize, flat: &[f64], k: usize) -> Scan {
    let o = i * 3;
    let (pl, pa, pb) = (points[o], points[o + 1], points[o + 2]);
    let mut s = Scan {
        best: 0,
        best_dist: f64::INFINITY,
        second_dist: f64::INFINITY,
    };
    for c in 0..k {
        let dl = pl - flat[c * 3];
        let da = pa - flat[c * 3 + 1];
        let db = pb - flat[c * 3 + 2];
        let d = dl * dl + da * da + db * db;
        if d < s.best_dist {
            s.second_dist = s.best_dist;
            s.best_dist = d;
            s.best = c;
        } else if d < s.second_dist {
            s.second_dist = d;
        }
    }
    s
}

/// `runLloydOnPoints`: assignments identical to a full scan every iteration; the bounds only skip provably
/// unnecessary scans.
fn run_lloyd(points: &[f64], initial: Vec<Oklab>) -> (Vec<Oklab>, Vec<u8>) {
    let n = points.len() / 3;
    let mut centroids = initial;
    let k = centroids.len();
    let mut assignments = vec![0u8; n];
    let mut flat = vec![0f64; k * 3];
    let mut sums = vec![0f64; k * 3];
    let mut counts = vec![0u32; k];
    let mut upper = vec![0f64; n];
    let mut lower = vec![0f64; n];
    let mut moved = vec![0f64; k];
    let mut half_gap = vec![0f64; k];

    let assign = |first: bool,
                  centroids: &[Oklab],
                  flat: &mut [f64],
                  half_gap: &mut [f64],
                  assignments: &mut [u8],
                  upper: &mut [f64],
                  lower: &mut [f64]| {
        for c in 0..k {
            flat[c * 3..c * 3 + 3].copy_from_slice(&centroids[c]);
        }
        for c in 0..k {
            let mut nearest = f64::INFINITY;
            for o in 0..k {
                if o == c {
                    continue;
                }
                let dl = flat[c * 3] - flat[o * 3];
                let da = flat[c * 3 + 1] - flat[o * 3 + 1];
                let db = flat[c * 3 + 2] - flat[o * 3 + 2];
                let d = dl * dl + da * da + db * db;
                if d < nearest {
                    nearest = d;
                }
            }
            half_gap[c] = nearest.sqrt() / 2.0;
        }
        let flat: &[f64] = flat;
        let half_gap: &[f64] = half_gap;
        assignments
            .par_chunks_mut(CHUNK)
            .zip(upper.par_chunks_mut(CHUNK))
            .zip(lower.par_chunks_mut(CHUNK))
            .enumerate()
            .for_each(|(chunk, ((assignments, upper), lower))| {
                for j in 0..assignments.len() {
                    let i = chunk * CHUNK + j;
                    if !first {
                        let a = assignments[j] as usize;
                        let bound = jsmath::max(lower[j], half_gap[a]);
                        if upper[j] * (1.0 + BOUND_MARGIN) < bound * (1.0 - BOUND_MARGIN) {
                            continue;
                        }
                        let s = a * 3;
                        let dl = points[i * 3] - flat[s];
                        let da = points[i * 3 + 1] - flat[s + 1];
                        let db = points[i * 3 + 2] - flat[s + 2];
                        upper[j] = (dl * dl + da * da + db * db).sqrt();
                        if upper[j] * (1.0 + BOUND_MARGIN) < bound * (1.0 - BOUND_MARGIN) {
                            continue;
                        }
                    }
                    let s = scan_point(points, i, flat, k);
                    assignments[j] = s.best as u8;
                    upper[j] = s.best_dist.sqrt();
                    lower[j] = s.second_dist.sqrt();
                }
            });
    };

    for iter in 0..MAX_ITERATIONS {
        assign(
            iter == 0,
            &centroids,
            &mut flat,
            &mut half_gap,
            &mut assignments,
            &mut upper,
            &mut lower,
        );

        sums.fill(0.0);
        counts.fill(0);
        for i in 0..n {
            let c = assignments[i] as usize;
            sums[c * 3] += points[i * 3];
            sums[c * 3 + 1] += points[i * 3 + 1];
            sums[c * 3 + 2] += points[i * 3 + 2];
            counts[c] += 1;
        }

        let mut max_shift_sq = 0.0;
        let previous = centroids.clone();
        for c in 0..k {
            if counts[c] == 0 {
                continue;
            }
            let cnt = counts[c] as f64;
            let next = [
                sums[c * 3] / cnt,
                sums[c * 3 + 1] / cnt,
                sums[c * 3 + 2] / cnt,
            ];
            max_shift_sq = jsmath::max(max_shift_sq, oklab_distance_sq(&centroids[c], &next));
            centroids[c] = next;
        }

        // moveBounds
        let mut farthest = 0.0;
        let mut farthest_centroid = usize::MAX;
        let mut second_farthest = 0.0;
        for c in 0..k {
            moved[c] = oklab_distance_sq(&previous[c], &centroids[c]).sqrt();
            if moved[c] > farthest {
                second_farthest = farthest;
                farthest = moved[c];
                farthest_centroid = c;
            } else if moved[c] > second_farthest {
                second_farthest = moved[c];
            }
        }
        let moved_now: &[f64] = &moved;
        assignments
            .par_chunks(CHUNK)
            .zip(upper.par_chunks_mut(CHUNK))
            .zip(lower.par_chunks_mut(CHUNK))
            .for_each(|((assignments, upper), lower)| {
                for j in 0..assignments.len() {
                    let a = assignments[j] as usize;
                    upper[j] += moved_now[a];
                    lower[j] -= if a == farthest_centroid {
                        second_farthest
                    } else {
                        farthest
                    };
                }
            });
        if max_shift_sq < CONVERGENCE_THRESHOLD_SQ {
            break;
        }
    }

    assign(
        false,
        &centroids,
        &mut flat,
        &mut half_gap,
        &mut assignments,
        &mut upper,
        &mut lower,
    );
    (centroids, assignments)
}

/// `buildPaletteFromAssignment`: empty clusters dropped.
fn build_palette(centroids: &[Oklab], assignments: &[u8]) -> (Vec<u8>, Vec<Rgb>) {
    let mut counts = vec![0usize; centroids.len()];
    for &c in assignments {
        counts[c as usize] += 1;
    }
    let mut remap = vec![0u8; centroids.len()];
    let mut palette = Vec::new();
    for (c, centroid) in centroids.iter().enumerate() {
        if counts[c] == 0 {
            continue;
        }
        remap[c] = palette.len() as u8;
        palette.push(oklab_to_rgb(*centroid));
    }
    (
        assignments.iter().map(|&c| remap[c as usize]).collect(),
        palette,
    )
}

fn inject_worst_fit(
    points: &[f64],
    assignment: &[u8],
    centroids: &[Oklab],
    slots: usize,
    importance: &[f32],
) -> Vec<Oklab> {
    let mut next_assignment = assignment.to_vec();
    let mut next_centroids = centroids.to_vec();
    let n = points.len() / 3;
    let mut assigned = vec![0f64; n];
    for i in 0..n {
        let c = next_centroids[next_assignment[i] as usize];
        let dl = points[i * 3] - c[0];
        let da = points[i * 3 + 1] - c[1];
        let db = points[i * 3 + 2] - c[2];
        assigned[i] = dl * dl + da * da + db * db;
    }
    for _ in 0..slots {
        // The first point with the largest score, as the sequential scan finds it.
        let (worst_score, worst) = (0..n)
            .into_par_iter()
            .with_min_len(CHUNK)
            .map(|i| {
                (
                    assigned[i] * (1.0 + WORST_FIT_IMPORTANCE_BOOST * importance[i] as f64),
                    i,
                )
            })
            .reduce(
                || (-1.0, usize::MAX),
                |a, b| {
                    if b.0 > a.0 || (b.0 == a.0 && b.1 < a.1) {
                        b
                    } else {
                        a
                    }
                },
            );
        let worst = if worst_score > -1.0 { worst } else { 0 };
        let new_centroid = point_at(points, worst);
        let new_index = next_centroids.len();
        next_centroids.push(new_centroid);
        let [cl, ca, cb] = new_centroid;
        next_assignment
            .par_chunks_mut(CHUNK)
            .zip(assigned.par_chunks_mut(CHUNK))
            .enumerate()
            .for_each(|(chunk, (next_assignment, assigned))| {
                for j in 0..assigned.len() {
                    let o = (chunk * CHUNK + j) * 3;
                    let dl = points[o] - cl;
                    let da = points[o + 1] - ca;
                    let db = points[o + 2] - cb;
                    let d = dl * dl + da * da + db * db;
                    if d < assigned[j] {
                        next_assignment[j] = new_index as u8;
                        assigned[j] = d;
                    }
                }
            });
    }
    next_centroids
}

fn plain_kmeans(points: &[f64], color_count: usize) -> (Vec<u8>, Vec<Rgb>) {
    let cell_count = points.len() / 3;
    let k = color_count.min(cell_count).max(1);
    let seed = (0xc0ffee_u32 as i32 ^ cell_count as i32 ^ k as i32) as u32;
    let mut rng = Mulberry32::new(seed);
    let seeds = kmeans_plus_plus_seeds(points, k, &mut rng);
    let (centroids, assignments) = run_lloyd(points, seeds);
    build_palette(&centroids, &assignments)
}

/// `quantizer.quantize(cells, colorCount, importance, cellOklab)` for the two built-in quantizers.
pub fn quantize(
    quantizer: Quantizer,
    points: &[f64],
    color_count: usize,
    importance: &[f32],
) -> (Vec<u8>, Vec<Rgb>) {
    let initial = plain_kmeans(points, color_count);
    if quantizer == Quantizer::Original {
        return initial;
    }
    let cell_count = points.len() / 3;
    let target_k = color_count.min(cell_count);
    if target_k < 3 {
        return initial;
    }
    let (merged_index, merged_palette) =
        merge_similar_colors(&initial.0, &initial.1, REINVEST_MERGE_THRESHOLD);
    let freed = target_k.saturating_sub(merged_palette.len());
    if freed == 0 {
        return initial;
    }
    let merged_oklab: Vec<Oklab> = merged_palette.iter().map(|&c| rgb_to_oklab(c)).collect();
    let centroids = inject_worst_fit(points, &merged_index, &merged_oklab, freed, importance);
    let (centroids, assignments) = run_lloyd(points, centroids);
    build_palette(&centroids, &assignments)
}
